apps/lib/apps.c

```diff
@@ -2720,3 +2721,30 @@ void app_params_free(OSSL_PARAM *params)
         OPENSSL_free(params);
     }
 }
+
+int app_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY **pkey, const char *alg, int bits)
+{
+    if (alg != NULL) {
+        if (bits > 0)
+            BIO_printf(bio_err, "Generating a %s key, with %d bits\n",
+                       alg, bits);
+        else
+            BIO_printf(bio_err, "Generating a %s key\n", alg);
+    }
+    if (!RAND_status())
+        BIO_printf(bio_err, "Warning: generating random key material may take long time if the system has a poor entropy source\n");
+    if (EVP_PKEY_keygen(ctx, pkey) > 0)
+        return 1;
+    BIO_puts(bio_err, "Generating private key failed\n");
+    return 0;
+}
+
+int app_paramgen(EVP_PKEY_CTX *ctx, EVP_PKEY **pkey)
+{
+    if (!RAND_status())
+        BIO_printf(bio_err, "Warning: ganerating random key parameters may take long time if the system has a poor entropy source\n");
+    if (EVP_PKEY_paramgen(ctx, pkey) > 0)
+        return 1;
+    BIO_puts(bio_err, "Generating key parameters failed\n");
```

Comment:
should these follow #11993 and even use app_bail_out?

Reply:
Good thought! Done.
This lead to saving a couple of code lines and error conditions.